### ISSUE TITLE
commit.in_main_branch returns True for commits that do not exist in the main branch.

### ISSUE THREAD

**Issue by gotec** [ishepard/pydriller#134]
**Describe the bug**
Dear all,

I have recently started mining branches (other than the main branch) with PyDriller. In doing so, I found that `commit.in_main_branch` also returns True for commits that do not exist in the main branch. Is this the expected behavior? If so, do you have another function that checks if a commit is present in the default branch?

Many thanks,
Christoph

**To Reproduce**
```
import pydriller
import git
local_dir = '.'
git_repo_dir_branch = 'tmp_git_branch'
git_repo_dir_master = 'tmp_git_master'

git.Git(local_dir).clone('https://github.com/gotec/git2net', git_repo_dir_master)
git_repo_master = pydriller.GitRepository(git_repo_dir_master)
commits_master = set(c.hash for c in git_repo_master.get_list_commits())

git.Git(local_dir).clone('https://github.com/gotec/git2net', git_repo_dir_branch, branch='mining_branches')
git_repo_branch = pydriller.GitRepository(git_repo_dir_branch)
commits_branch = set(c.hash for c in git_repo_branch.get_list_commits())

commits_only_in_branch = list(commits_branch.difference(commits_master))

commit = git_repo_branch.get_commit(commits_only_in_branch[0])
print(commits_only_in_branch[0], commit.in_main_branch)
```

**OS Version:**
Linux

**Comment by ishepard** [comment]
Hi @gotec!
I haven't run your code, but I wanted to explain a bit this "main_branch" thing.
Git doesn't really have the concept of main branch, indeed you can rename the "master" branch to whatever name you'd like.
However, there is a default branch, that is the branch that is checked out when you first clone the repository.

Hence, in Pydriller the main branch is the branch that is checked out when you run the study, that should be the default. However, if you first checkout a different branch B1 and then run Pydriller, it will say that B1 is the main branch. 
TBH, I even thought to remove this "main_branch" thing because it's not really helpful, I don't think anyone actually uses it.

**Comment by ishepard** [comment]
I had a look at your code and indeed it's what I was thinking. 
By itself the result is correct, it says that `['f58d707088feea78fae16aedd5bc70d38b0fe6fd', '089ec3abb570f10df47de38af129366a314be9f7']` are the 2 commits that are in the `mining_branches` branch.

However, since your default branch is `mining_branches`, Pydriller outputs that they are in the main branch.
This code instead will work:
```
import pydriller
import git
local_dir = '.'
git_repo_dir_branch = 'tmp_git_branch'
git_repo_dir_master = 'tmp_git_master'

git.Git(local_dir).clone('https://github.com/gotec/git2net', git_repo_dir_master)
git_repo_master = pydriller.GitRepository(git_repo_dir_master)
commits_master = set(c.hash for c in git_repo_master.get_list_commits())

git.Git(local_dir).clone('https://github.com/gotec/git2net', git_repo_dir_branch, branch='mining_branches')
git_repo_branch = pydriller.GitRepository(git_repo_dir_branch)
commits_branch = set(c.hash for c in git_repo_branch.get_list_commits())

commits_only_in_branch = list(commits_branch.difference(commits_master))

for commit in commits_only_in_branch:
    c = git_repo_master.get_commit(commit)
    print(c.in_main_branch)
```

```
False
False
```

**Comment by gotec** [comment]
Thanks for the super quick answer! I am aware of the idea of the "default" branch  (which is usually the "master" branch). However, I was expecting the "default" branch to be the one that I check out when I do not provide a specific branch (see the first clone). However, I now understand your implementation which specifies that a commit is not in the branch that is currently checked out.

Thanks for clarifying this, in this case I might have to rethink my current mining process or drop that entry too. Is there any PyDriller function that lists all the commits in the repository? `get_list_commits()` seems to only list the commits in the current branch.

Best,
Christoph

**Comment by ishepard** [comment]
I think what you are looking for is something like "git log --all", that includes the commits starting from all the refs. In Pydriller this is implemented as "include_refs=True".

**Comment by gotec** [comment]
Thanks a lot, I will try it out!

Cheers,
Christoph

**Comment by gotec** [comment]
Seems like `git_repo.get_list_commits(all=True)` is exactly what I was looking for. Thanks again for the clarifications. I consider my issue resolved.

Best,
Christoph

**Comment by ishepard** [comment]
Indeed I was able to reproduce. Though it's not Pydriller fault :) 
If you navigate to `tmp_git_branch` (or `tmp_git_master`) and you get the branches of commit `40e36389f3443c8feb2a9274104c85e7e9c57fd5`:
```
git branch --contains 40e36389f3443c8feb2a9274104c85e7e9c57fd5
```
you will see:
```
* mining_branches
(END)
```
so this commit (according to git) is only on that branch. Of course it's not right.
So try to obtain the list of branches:
```
git branch --list
```
and you'll get:
```
* mining_branches
(END)
```
So apparently, there is only one branch in your local copy. That's because the way you cloned the repo.
If you type:
```
git branch --list -r
```
`-r` stands for "give me also the remote branches", you'll get:
```
  origin/HEAD -> origin/master
  origin/add-code-of-conduct-1
  origin/master
  origin/mining_branches
(END)
```
so when cloning the repo, you didn't clone all the branches but only the specific one.

**Comment by gotec** [comment]
This is interesting! Do you have any idea why I can then get a commit from the repository (see your solution 
from before https://github.com/ishepard/pydriller/issues/134#issuecomment-667707651) if that commit doesn't actually exist in the repository?

Thanks also for the quick fix that I saw while typing this!

**Comment by ishepard** [comment]
Mmmm not sure about this one, I'd have to google a bit to find out 😄 but from my understanding, you already have all the commits in your local repo, but if you don't checkout the branch it's not indexed. But it's there anyway.
But this is just my guess, I am not sure..git works in mysterious ways 😄

**Comment by gotec** [comment]
Yes, I have found that out the hard way trying to account for way to many special corner cases over the past year.

Anyhow, thanks a lot for sharing all of this. Based on what I learned I will ensure to "properly" clone repositories in the future and likely stop using the "in_main_branch" indicator as that yields different results based on the branch that is currently checked out. I still don't find this behaviour very intuitive but as you so rightly pointed out both are not issues with pyDriller but rather with git itself (and hence nothing we can change).

In case you find out anything regarding the commits not in the repo (see https://github.com/ishepard/pydriller/issues/134#issuecomment-668680327) I would still be very interested to learn about this. Otherwise feel free to close this issue.

Again thanks a lot for all the information!

Best,
Christoph
